cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
     return String.format("%s:%s:%s:%s", program.getNamespaceId(), program.getApplicationId(),
                          programType.name(), program.getId());
   }
+
+  /**
+   * One instance of this class contains a list of {@link StreamSizeSchedule}s, which are all interested
+   * in the same stream. This instance subscribes to the size notification of the stream, and polls the
+   * stream for its size whenever the schedules it references need the information.
+   * The {@link StreamSizeScheduler} communicates with this class, which in turn communicates to the schedules
+   * it contains to perform operations on the schedules - suspend, resume, etc.
+   */
+  private final class StreamSubscriber implements NotificationHandler<StreamSizeNotification>, Cancellable {
+    // Key is the schedule ID
+    private final ConcurrentMap<String, StreamSizeScheduleTask> scheduleTasks;
+    private final Object lastNotificationLock;
+    private final Id.Stream streamId;
+    private final AtomicInteger activeTasks;
+
+    private Cancellable notificationSubscription;
+    private ScheduledFuture<?> scheduledPolling;
+    private StreamSizeNotification lastNotification;
+
+    private StreamSubscriber(Id.Stream streamId) {
+      this.streamId = streamId;
+      this.scheduleTasks = Maps.newConcurrentMap();
+      this.lastNotificationLock = new Object();
+      this.activeTasks = new AtomicInteger(0);
+    }
+
+    public void start() throws NotificationFeedException, NotificationFeedNotFoundException {
```

Comment:
If it succeeds, I am not sure if the subscription can be interrupted. The notification service relies on Twill's ``KafkaClient`` to subscribe to the notifications in Kafka. I don't know how Twill handles possible interruption in the subscription, or if it is even possible. @chtyim, what is your take on this?

It could fail for 2 reasons: 
- if the stream does not exist, then a NotificationFeedNotFoundException will be thrown. I think it's a legitimate error, which should revert the creation of the schedule. When deploying an app that has one such schedule, the exception should be caught and logged, but the app should still be deployed.
- if the notification service cannot run, for eg because the dataset service is not up, then a NotificationFeedException will be thrown. In that case we should retry, you're right. But what should define how long we retry?